### PR TITLE
Advanced HW7 - 'gcc flags' - adawes

### LINKED ISSUE #229: Creating list of useful compiling Flags

### PR DESCRIPTION
**Pull request by anthonydawes**
# Advanced HW 7 submission

Closes #229

Name: Anthony Dawes
Uniqname: adawes

### CODE REVIEW

**Comment by mterwill** [comment]
Please fill out the PR template.

File: _commands/basics/gcc.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
+---
+---
+
+gcc flags
+--
+
+`gcc` is a compiler command that has the ability to execute options or 'Flags' when compiling.  These flags can be grouped and executed via the command line:
+    '-o' places output in a file.
```
Comment (mterwill) [inline _commands/basics/gcc.md]: For code snippets, generally prefer \`backticks\` over 'single quotes' so they `render nicely`.

Additionally, these all collapse into one ¶ when rendered:

image

Take a look at [how to format with Markdown](https://daringfireball.net/projects/markdown/).

File: _commands/basics/gcc.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+### Useful Options / Examples
+
+#### `-o`
+~~~ bash
+$ gcc test.c -o test
+~~~
+
+##### Break it down
+~~~ bash
+*  This is a very simple example where test.c is being compiled and outputted to the test file.  The output filename is specified after the flag.  Input: test.c, Output: test
```
Comment (mterwill) [inline _commands/basics/gcc.md]: Should these be code blocks?

image

Comment (anthonydawes) [inline _commands/basics/gcc.md]: hi Matt, I am not sure if I am suppose to do something about this?  Could you please, advise?

File: _commands/basics/gcc.md (hunk 1 above)
Comment (anthonydawes) [inline _commands/basics/gcc.md]: Thank You!  I will try to fix it.

File: _commands/basics/gcc.md (hunk 2 above)
Comment (mterwill) [inline _commands/basics/gcc.md]: Yes, please delete the code blocks surrounding the "Break it down" section.

File: _commands/basics/gcc.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+
+#### `-ar`
+~~~ bash
+$ gcc -ar test.c
+~~~
+
+##### Break it down
+
+This option creates an archive (a static library) instead of a program. The resulting file will have an .a ending.
```
Comment (mterwill) [inline _commands/basics/gcc.md]: I'm unfamiliar with this – could you maybe expand a bit on the implications here?

File: _commands/basics/gcc.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+`-ar` creates an archive instead of a program.
+
+~~~ bash
+$ gcc test.c -o test
+$ gcc -g test.c
+$ gcc -Werror test.c
+$ gcc -Wall test.c
+$ gcc -ar test.c
+~~~
```
Comment (mterwill) [inline _commands/basics/gcc.md]: I think just one example here of a common use case should be sufficient.

File: _commands/basics/gcc.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
+---
+---
+
+gcc flags
+--
+
+`gcc` is a compiler command that has the ability to execute options or 'Flags' when compiling.  These flags can be grouped and executed via the command line:
+
+`-o` places output in a file.
```
Comment (mterwill) [inline _commands/basics/gcc.md]: Could you please elaborate on what happens if you do _not_ pass the `-o` flag?

**Comment by anthonydawes** [comment]
I believe all of the comments are taken care of now

File: _commands/basics/gcc.md (hunk 5 above)
Comment (anthonydawes) [inline _commands/basics/gcc.md]: Yes, I added the explanation. TY!

File: _commands/basics/gcc.md (hunk 4 above)
Comment (anthonydawes) [inline _commands/basics/gcc.md]: Yes, great call.  I ended up switching the `-ar` flag to the `-O` flag.  The `-ar` was problematic. I thought the optimization was more straightforward and useful.  So, I combined the statements, like you commented.

File: _commands/basics/gcc.md (hunk 3 above)
Comment (anthonydawes) [inline _commands/basics/gcc.md]: I mentioned this in my above comment, but my understand was that the `-ar` flag was a local copy or archive for records (somewhat of a local git), but this was problematic.  I switched that flag for the optimization flag which I believe is quite a bit more useful.  Thank You for all of your comments!  I appreciate your time!!

File: _commands/basics/gcc.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+~~~ bash
+$ gcc test.c -o test
+$ gcc -g test.c
+$ gcc -Werror -Wall -O test.c
+~~~
+
+
+
+<!--more-->
```
Comment (mterwill) [inline _commands/basics/gcc.md]: Let's put this closer to the top so we don't have so much showing on the index page.

File: _commands/basics/gcc.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
+---
+---
+
+gcc flags
+--
+
+`gcc` is a compiler command that has the ability to execute options or 'Flags' when compiling.  These flags can be grouped and executed via the command line:
+
+`-o` places output in a filename of your choice.  If you do not use the `-o` flag the output file will have the same name as the file being compiled without any extension.
```
Comment (mterwill) [inline _commands/basics/gcc.md]: Not `a.out`?

File: _commands/basics/gcc.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+
+<!--more-->
+
+### Useful Options / Examples
+
+#### `-o`
+~~~ bash
+$ gcc test.c -o test
+$ gcc test.c -o
```
Comment (mterwill) [inline _commands/basics/gcc.md]: This (`gcc test.c -o`) is a little confusing; let's just get rid of it.

File: _commands/basics/gcc.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+In the first line of code, the `-O` option instructs the compiler to try and reduce code size and execution time, without performing any optimizations that take a great deal of compilation time.  There are additional variations to the `-O` option that use alternative methods to optimize a file.  Examples include:  `-O1`, `-O2`, `-O3` and `-Os`.
+
+
+`-O1` Further Optimizes where the compilation takes somewhat more time, and a lot more memory for a large function.
+
+
+`-O2` Optimizes even more. GCC performs nearly all supported optimizations that do not involve a space-speed tradeoff.
+
+
+`-O3` Optimize yet more. -O3 turns on all optimizations specified by -O2 and also turns on the -finline-functions, -funswitch-loops, -fpredictive-commoning, -fgcse-after-reload, -ftree-loop-vectorize, -ftree-loop-distribute-patterns, -fsplit-paths -ftree-slp-vectorize, -fvect-cost-model, -ftree-partial-pre, -fpeel-loops and -fipa-cp-clone options.
```
Comment (mterwill) [inline _commands/basics/gcc.md]: Can you put these all in `code snippets`? (use \`backticks\`)

The flag references are also a little inconsistently styled (-O3 some places and `-O3` in others)

File: _commands/basics/gcc.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+
+
+
+
+#### `-O`
+~~~ bash
+$ gcc -O test.c
+$ gcc -O1 test.c
+
```
Comment (mterwill) [inline _commands/basics/gcc.md]: Extra newline here.

File: _commands/basics/gcc.md [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+`-o` places output in a filename of your choice.  If you do not use the `-o` flag the output file will have the same name as the file being compiled without any extension.
+
+`-g` turns on debugging information. 
+
+`-Werror` makes all warnings into errors.
+
+`-Wall` enables all the warnings.
+
+`-O` enables optimization.
```
Comment (mterwill) [inline _commands/basics/gcc.md]: Could you please write a note to see below for details?

File: _commands/basics/gcc.md (hunk 7 above)
Comment (anthonydawes) [inline _commands/basics/gcc.md]: Thank You!  I was going off of memory... sorry!   This change is done and I relearned it :)

File: _commands/basics/gcc.md (hunk 11 above)
Comment (anthonydawes) [inline _commands/basics/gcc.md]: Yes, np, note inserted.

File: _commands/basics/gcc.md (hunk 8 above)
Comment (anthonydawes) [inline _commands/basics/gcc.md]: Deleted.

File: _commands/basics/gcc.md (hunk 10 above)
Comment (anthonydawes) [inline _commands/basics/gcc.md]: newline deleted.

File: _commands/basics/gcc.md (hunk 9 above)
Comment (anthonydawes) [inline _commands/basics/gcc.md]: Done. great catch. and np... if I knew better what i was doing, I would have gotten it the first time.  Thank You for you time and patience!

**Comment by anthonydawes** [comment]
HI Matt, I believe I addressed all of your comments.  TY

**Comment by mterwill** [comment]
Looks great! Thanks for your work on this :smile:

Merging + giving you credit.

**Comment by anthonydawes** [comment]
Thank You, Matt!  Great working with you on this.  I learned a lot 😀
Anthony

**Pull request merged**